File: ios/ExpoPasskeyLivenessModule.swift
import ExpoModulesCore
import AVFoundation

public class ExpoPasskeyLivenessModule: Module {
  /// Cancellation handle for the in-flight adapter run, if any.
  private var cancelHandler: (() -> Void)?

  public func definition() -> ModuleDefinition {
    Name("ExpoPasskeyLivenessModule")

    Function("isLivenessSupported") { () -> Bool in
      return AVCaptureDevice.default(
        .builtInWideAngleCamera,
        for: .video,
        position: .front
      ) != nil
    }

    AsyncFunction("runLivenessCheck") { (options: [String: Any], promise: Promise) in
      guard let providerName = options["provider"] as? String else {
        promise.reject(
          "liveness_invalid_bootstrap",
          "runLivenessCheck: 'provider' is required"
        )
        return
      }
      guard let bootstrapStr = options["bootstrap"] as? String,
            let bootstrapData = bootstrapStr.data(using: .utf8),
            let bootstrap = try? JSONSerialization.jsonObject(with: bootstrapData) as? [String: Any] else {
        promise.reject(
          "liveness_invalid_bootstrap",
          "runLivenessCheck: 'bootstrap' is not valid JSON"
        )
        return
      }

      guard let adapter = LivenessAdapterRegistry.adapter(forProviderName: providerName) else {
        promise.reject(
          "liveness_not_supported",
          "No adapter registered for provider \(providerName). Add it to the expo-passkey-liveness config plugin providers list and re-run prebuild."
        )
        return
      }

      let timeoutMs = options["timeoutMs"] as? Int
      let locale = options["locale"] as? String

      // Pre-flight: explicit camera permission check, since the
      // Amplify pod will prompt internally but we want a stable code
      // when permission is already denied.
      let status = AVCaptureDevice.authorizationStatus(for: .video)
      if status == .denied || status == .restricted {
        promise.reject("liveness_camera_permission_denied", "Camera permission denied")
        return
      }

      Task {
        do {
          let payload = try await adapter.run(
            bootstrap: bootstrap,
            timeoutMs: timeoutMs,
            locale: locale
          )
          promise.resolve(payload)
        } catch let err as LivenessException {
          // err.reason carries "code::message"
          let parts = err.reason.split(separator: ":", maxSplits: 1).map(String.init)
          let code = parts.first ?? "liveness_provider_error"
          let message = parts.count > 1 ? parts[1].trimmingCharacters(in: CharacterSet(charactersIn: ":")) : err.reason
          promise.reject(code, message)
        } catch {
          promise.reject("liveness_provider_error", error.localizedDescription)
        }
      }
    }

    AsyncFunction("cancel") { (promise: Promise) in
      self.cancelHandler?()
      self.cancelHandler = nil
      promise.resolve(nil)
    }
  }
}

File: android/src/main/java/expo/modules/passkeyliveness/ExpoPasskeyLivenessModule.kt
package expo.modules.passkeyliveness

import android.Manifest
import android.app.Activity
import android.content.pm.PackageManager
import androidx.core.content.ContextCompat
import expo.modules.kotlin.modules.Module
import expo.modules.kotlin.modules.ModuleDefinition
import expo.modules.passkeyliveness.adapters.LivenessAdapterRegistry
import kotlinx.coroutines.CoroutineScope
import kotlinx.coroutines.Dispatchers
import kotlinx.coroutines.launch
import org.json.JSONObject

class ExpoPasskeyLivenessModule : Module() {

  private val scope = CoroutineScope(Dispatchers.Main)

  override fun definition() = ModuleDefinition {
    Name("ExpoPasskeyLivenessModule")

    Function("isLivenessSupported") {
      val ctx = appContext.reactContext ?: return@Function false
      ctx.packageManager.hasSystemFeature(PackageManager.FEATURE_CAMERA_FRONT)
    }

    AsyncFunction("runLivenessCheck") { options: Map<String, Any?>, promise: expo.modules.kotlin.Promise ->
      val providerName = options["provider"] as? String
      if (providerName.isNullOrEmpty()) {
        promise.reject(LivenessErrorCodes.INVALID_BOOTSTRAP, "Missing 'provider'", null)
        return@AsyncFunction
      }
      val bootstrapStr = options["bootstrap"] as? String
      if (bootstrapStr.isNullOrEmpty()) {
        promise.reject(LivenessErrorCodes.INVALID_BOOTSTRAP, "Missing 'bootstrap'", null)
        return@AsyncFunction
      }
      val bootstrap = try {
        JSONObject(bootstrapStr)
      } catch (err: Throwable) {
        promise.reject(LivenessErrorCodes.INVALID_BOOTSTRAP, "Bootstrap is not valid JSON: ${err.message}", err)
        return@AsyncFunction
      }
      val adapter = LivenessAdapterRegistry.adapter(providerName)
      if (adapter == null) {
        promise.reject(
          LivenessErrorCodes.NOT_SUPPORTED,
          "No adapter registered for provider $providerName. Add it to the expo-passkey-liveness config plugin providers list and rebuild.",
          null
        )
        return@AsyncFunction
      }

      val activity: Activity = appContext.currentActivity ?: run {
        promise.reject(
          LivenessErrorCodes.PROVIDER_ERROR,
          "No foreground activity available to host the liveness check",
          null
        )
        return@AsyncFunction
      }

      // Pre-flight: stable error when camera permission is already denied.
      val cameraGranted = ContextCompat.checkSelfPermission(activity, Manifest.permission.CAMERA) ==
        PackageManager.PERMISSION_GRANTED
      if (!cameraGranted) {
        promise.reject(
          LivenessErrorCodes.CAMERA_PERMISSION_DENIED,
          "Camera permission denied",
          null
        )
        return@AsyncFunction
      }

      val timeoutMs = (options["timeoutMs"] as? Number)?.toInt()
      val locale = options["locale"] as? String

      scope.launch {
        try {
          val payload = adapter.run(activity, bootstrap, timeoutMs, locale)
          promise.resolve(payload)
        } catch (err: LivenessException) {
          promise.reject(err.errorCode, err.message ?: err.errorCode, err)
        } catch (err: Throwable) {
          promise.reject(LivenessErrorCodes.PROVIDER_ERROR, err.message ?: err.toString(), err)
        }
      }
    }

    AsyncFunction("cancel") { promise: expo.modules.kotlin.Promise ->
      // No global cancel handle on Android yet — adapters resolve when
      // their activities finish, including when the user backs out.
      promise.resolve(null)
    }
  }
}
